enhancements/ota/distribute-secondary-metadata-as-container-image.md

```diff
@@ -0,0 +1,165 @@
+
+#### Mitigation
+A mitigation strategy is to sign the resulting image in the CPaaS pipeline, publish the signatures, and have Cincinnati verify the image after download.
+
+### cincinnati-graph-data source repository integrity
+Not strictly related to this enhancement, is the fact that we are not signing the commits in the source repository in the first place.
+By this we effectively trust GitHub not to change the content.
+
+#### Mitigation
+Enforce signed commits in the source repository.
+
+## Design Details
+
+### Test Plan
+**Note:** *Section not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+
+### Graduation Criteria
+**Note:** *Section not required until targeted at a release.*
+
+- Maturity levels - `Developer Preview`, `Red Hat Staging`, `Red Hat Production`, `GA`
+
+##### `Developer Preview`
+- New Cincinnati plugin
+- Container image registry managed by and dedicated to prow CI
+- Integration tests using an image (to-be-determined: verifying against mocked signatures)
+- CPaaS pipeline is setup up to
+
+##### `Developer Preview` -> `Red Hat Staging`
+- Update the Cincinnati deployment template to use the new plugin.
+
+##### Red Hat Staging -> Red Hat Production
+- New plugin is successfully deployed to production
+- Adjust the Cincinnati Operator to use the new plugin
+
+##### Red Hat Production -> GA
+- Document the mirroring process for end-users
+- Definitive positive feedback from the ACM team and their use-case
+
+##### Removing a deprecated feature
+*Not applicable*
+
+### Upgrade / Downgrade Strategy
+*Not applicable*
+
+### Version Skew Strategy
+*Not applicable*
+
+## Implementation History
+
+### Initial implementation
+The initial implementation involves setting up the CPaaS pipeline, implementing a new Cincinnati plugin which downloads the published image and extracts it to disk for subsequent processing in the plugin chain.
+It also involves adjusting the Cincinnati Operator to use the new plugin by default.
+
+## Drawbacks
+
+### Increase OCP release latency
+The enhancement introduces the additional latency of the CPaaS pipeline into the OCP release workflow.
+Assuming that the pipeline is fully automated, the latency introduced is expected to be several orders of magnitude lower (minutes) compared to the OCP release cadence (hours-days).
+
+## Alternatives
+*None, other than leaving things as they are.*
```

Comment:
Shipping secondary-metadata into the cluster via an alternative mechanism (tarball, ConfigMap, custom resource, whatever), and having the local Cincinnati operator provide a mechanism for the admin to submit it for injection into the Cincinnati container (ConfigMap, URI pointing at a tarball on some internal HTTPS host, local Git host, whatever).  Upside of those approaches is that they don't involve CPaaS, its latency, and image signing (which is not well standardized), and instead rely on transitive trust (which is also not well standardized, but which is up to the user and does not need to be explained to Cincinnati or its operator).  Downside is that you need a channel for pushing that data into restricted-network clusters, and #188 (which was setting itself up to be that channel) was rejected in favor of the narrower #283 (which explicitly excludes itself from being that channel).

Reply:
Note that this enhancement is not only about the network-restricted operation of Cincinnati. So I'm not sure which of the things you're mentioning is an alternative on the level of this enhancement. I see them as alternatives to replace specific aspects of it.

Reply:
I tried to extract the essence of your comment into the `alternatives` section.